build/pluginctl/logs.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/mattermost/mattermost/server/public/model"
+)
+
+const (
+	logsPerPage     = 100 // logsPerPage is the number of log entries to fetch per API call
+	timeStampFormat = "2006-01-02 15:04:05.000 Z07:00"
+)
+
+// logs fetches the latest 500 log entries from Mattermost,
+// and prints only the ones related to the plugin to stdout.
+func logs(ctx context.Context, client *model.Client4, pluginID string) error {
+	for i := 5 - 1; i >= 0; i-- {
+		logs, err := fetchLogs(ctx, client, i, pluginID, time.Unix(0, 0))
+		if err != nil {
+			return fmt.Errorf("failed to fetch log entries: %w", err)
+		}
+
+		err = printLogEntries(logs)
+		if err != nil {
+			return fmt.Errorf("failed to print logs entries: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// watchLogs fetches log entries from Mattermost and print them to stdout.
+// It will return without an error when ctx is canceled.
+func watchLogs(ctx context.Context, client *model.Client4, pluginID string) error {
+	now := time.Now()
+	var oldest string
+
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			for {
+				logs, err := fetchLogs(ctx, client, 0, pluginID, now)
+				if err != nil {
+					return fmt.Errorf("failed to fetch log entries: %w", err)
+				}
+
+				var allNew bool
+				logs, oldest, allNew = checkOldest(logs, oldest)
+
+				err = printLogEntries(logs)
+				if err != nil {
+					return fmt.Errorf("failed to print logs entries: %w", err)
+				}
+
+				if !allNew {
+					// No more logs to fetch
+					break
+				}
```

Comment:
I don't understand this `allNew` check. If we get a set of all new logs, we loop to make exactly the same `fetchLogs` call. If no new logs have been emitted, we'll get exactly the same payload, ignore all the logs, and then `break`. If new logs have been emitted, we'll spit those out (possibly looping again if they were all new)... but wouldn't we have just seen those logs on the next ticker event anyway?

I'm wondering if we can simplify to just fetching the single page without a loop.

But

Reply:
There was indeed a bug in the code. The correct behaviour is to fetch the next page of log entries if the previous one contained only new entries.

Fixed via 24a84b775adf6ccdcff72e4b0ce74954b1203210.